only show directly inherited types in the InheritedBy progression query

brettfo opened the pull request:
Fixes internal bug 747965.

When showing the derived types in a progression query in the Solution Explorer, only the types that immediately derive from the selected type should be shown, not everything down the chain (because the tree view will take care of that).  This takes the behavior back to how it was in Dev12.

brettfo commented:
Tagging @Pilchie @dpoeschl @jasonmalinowski @DustinCampbell @rchande @balajikris @jmarolf @davkean as potential reviewers.

Pilchie commented on src/VisualStudio/Core/Def/Implementation/Progression/GraphQueries/InheritedByGraphQuery.cs:
```diff
@@ -32,7 +27,7 @@ internal sealed class InheritedByGraphQuery : IGraphQuery
                 if (namedType.TypeKind == TypeKind.Class)
                 {
                     var derivedTypes = await namedType.FindDerivedClassesAsync(solution, cancellationToken).ConfigureAwait(false);
```
Doesn't this does a bunch of extra work to go and find all the levels of inheritance, only to ignore it after you have it?  Consider making `FindDerivedClassesAsync` able to return just one level directly.

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/Progression/GraphQueries/InheritedByGraphQuery.cs:
```diff
@@ -32,7 +27,7 @@ internal sealed class InheritedByGraphQuery : IGraphQuery
                 {
                     var derivedTypes = await namedType.FindDerivedClassesAsync(solution, cancellationToken).ConfigureAwait(false);
-                    foreach (var derivedType in derivedTypes)
+                    foreach (var derivedType in derivedTypes.Where(derived => derived.BaseType == namedType))
```
We don't have another helper for this somewhere? I feel bad we're doing lots of work chasing types only to throw them out.

dpoeschl commented on src/VisualStudio/Core/Def/Implementation/Progression/GraphQueries/InheritedByGraphQuery.cs:
```diff
@@ -41,7 +36,7 @@ internal sealed class InheritedByGraphQuery : IGraphQuery
                 else if (namedType.TypeKind == TypeKind.Interface)
                 {
                     var derivedTypes = await namedType.FindDerivedInterfacesAsync(solution, cancellationToken).ConfigureAwait(false);
```
Unrelated to your change, but in Dev12 the Derived Types query we used to include classes that implemented the interface, but in the same hierarchical fashion that you're doing here now (as opposed to the new Implemented By query where flatten the list of implementers). Let's consider bringing that back.

brettfo commented:
I've updated the helpers (only called by this progression query) to filter types when searching and not after, so we avoid doing extra work.  I've also added more to the tests that verify we brought the behavior back in line with Dev12 as mentioned in @dpoeschl's comment around showing types that directly implement an interface.

Pilchie commented:
:+1:

dpoeschl commented on src/VisualStudio/Core/Test/Progression/InheritedByGraphQueryTests.vb:
```diff
@@ -74,10 +78,16 @@ using System;
 
 public interface $$I { }
 
+public class C : I { } // should appear as being derived from (implementing) I
```
Do you have/need a case for basically ```class C2 : C {} // should not appear as being derived from I```?

brettfo replied:
Good catch, I'll add that.

jasonmalinowski commented on src/Workspaces/Core/Portable/Shared/Extensions/INamedTypeSymbolExtensions.cs:
```diff
@@ -25,12 +25,12 @@ public static IEnumerable<INamedTypeSymbol> GetBaseTypesAndThis(this INamedTypeS
             CancellationToken cancellationToken)
         {
-            return FindDerivedClassesAsync(type, solution, null, cancellationToken);
+            return DependentTypeFinder.GetTypesImmediatelyDerivedFromClassesAsync(type, solution, cancellationToken);
```
Nitpick: can we just get rid of extension methods like this that just forward to a static somewhere? It just makes the codebase harder to navigate.

jasonmalinowski commented on src/VisualStudio/Core/Test/Progression/InheritedByGraphQueryTests.vb:
```diff
@@ -74,10 +78,18 @@ using System;
     void M();
 }
+
+interface I3 : I2 // should not be shown as inherited by I
```
Can we have a VB test for these as well? VB's syntax at least treats these cases as different, I forget if the symbol model does.

jasonmalinowski commented:
:+1:, other than the build failing and nitpicks. :-)

brettfo commented:
Added VB tests and removed single-use extension methods to directly call the static helpers instead.
